File: 2452-words-within-two-edits-of-dictionary/solution.py
class Solution:
    def wordsWithinTwoEdits(self, queries: list[str], dictionary: list[str]) -> list[str]:
        ans = []
        
        for query in queries:
            # Check this query against every word in the dictionary
            for word in dictionary:
                diff_count = 0
                # All words have same length, so we can iterate by index
                for char_q, char_w in zip(query, word):
                    if char_q != char_w:
                        diff_count += 1
                    
                    # Optimization: stop comparing if we exceed 2 edits
                    if diff_count > 2:
                        break
                
                # If the Hamming distance is 2 or less, the query is valid
                if diff_count <= 2:
                    ans.append(query)
                    break
                    
        return ans

File: 2452-words-within-two-edits-of-dictionary/solution.cpp
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
    vector<string> wordsWithinTwoEdits(vector<string>& queries, vector<string>& dictionary) {
        vector<string> result;
        
        for (const string& query : queries) {
            bool foundMatch = false;
            
            for (const string& word : dictionary) {
                int diffCount = 0;
                
                // Compare query and word character by character
                for (int i = 0; i < query.length(); ++i) {
                    if (query[i] != word[i]) {
                        diffCount++;
                    }
                    // Optimization: If differences exceed 2, this word is not a match
                    if (diffCount > 2) break;
                }
                
                // If we found a word in the dictionary within 2 edits
                if (diffCount <= 2) {
                    foundMatch = true;
                    break;
                }
            }
            
            if (foundMatch) {
                result.push_back(query);
            }
        }
        
        return result;
    }
};